csrc/scheduler/registry_utils.cpp

```diff
@@ -572,59 +572,96 @@ bool SchedulerTopologyChecker::hasNonNormalizePostReductionBCast(
-        // in the pair will be the original id so we can reset it if it's not
-        // resolved before the next traversal. The second ID will be
-        // propagated as we map the IDs through the backward traversal.
-        std::vector<std::pair<IterDomain*, IterDomain*>> ids_to_resolve;
-
-        // Check if any TensorViews have a resolved broadcast
-        for (auto entry : forward_p2c_root_map) {
-          auto p_id = entry.first;
-          auto c_id = entry.second;
-          if (p_id->isBroadcast() && !c_id->isBroadcast()) {
-            ids_to_resolve.emplace_back(c_id, c_id);
+          // Check if any TensorViews have a resolved broadcast
+          for (auto entry : forward_p2c_root_map) {
+            auto p_id = entry.first;
+            auto c_id = entry.second;
+            if (p_id->isBroadcast() && !c_id->isBroadcast()) {
+              ids_to_resolve.emplace_back(c_id, c_id);
+            }
+          }
+          if (ids_to_resolve.empty()) {
+            continue;
           }
         }
 
-        if (ids_to_resolve.empty()) {
-          continue;
+        // At this point, we have some ids to resovle.
+        // Check if we can move forward further along the use chain.
+        // This helps gathering more usage info to resolve this id to avoid
+        // unnecessary segmentation when the reduction input iteration domain is
+        // used after this tensor. See PostReductionBroadcastCheck test and
+        // issue-2146. We can move forward when:
+        // (1) The usage chain is not empty.
+        // (2) The consumer is not a broadcastOp, as broadcastOp introduces a
+        //     new broadcast axis and require a new check, see
+        //     FusionReshapePersistentShmoo.
+        // (3) At least one id is mapped, otherwise can't do backward traversal.
+        if (!forward_tv_dep_chain.empty()) {
+          auto tmp_producer = forward_running_consumer;
+          auto tmp_consumer = forward_tv_dep_chain.front();
+          if (!tmp_consumer->definition()->isA<BroadcastOp>()) {
+            auto forward_pairwise_root_map =
+                PairwiseRootDomainMap(tmp_producer, tmp_consumer);
+            auto forward_p2c_root_map =
+                forward_pairwise_root_map.mapProducerToConsumer();
+            bool at_leat_one_id_mapped = false;
+            for (size_t entry_i = ids_to_resolve.size(); entry_i > 0;
+                 entry_i--) {
+              auto running_id = ids_to_resolve[entry_i - 1].second;
+              if (forward_p2c_root_map.find(running_id) !=
+                  forward_p2c_root_map.end()) {
+                at_leat_one_id_mapped = true;
+                ids_to_resolve[entry_i - 1] = std::make_pair(
+                    forward_p2c_root_map.at(running_id),
+                    forward_p2c_root_map.at(running_id));
+              }
+            }
+            if (at_leat_one_id_mapped) {
+              continue;
```

Comment:
Can't this continue even if nothing is mapped?

Reply:
we can't, otherwise there is no way to transverse back if there is no other links. For example:
![image](https://github.com/NVIDIA/Fuser/assets/116412316/90c8d7c3-5519-4da0-ba8d-c0271e76aa87)

Reply:
Thanks. Makes sense.